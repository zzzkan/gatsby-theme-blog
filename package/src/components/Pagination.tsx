import React from "react"
import { Link as GatsbyLink } from "gatsby"
import { Box, HStack, Link } from "@chakra-ui/react"

export type PaginationProps = {
  readonly basePath: string
  readonly totalPage: number
  readonly currentPage: number
}

const width = 2

export const Pagination: React.FC<PaginationProps> = ({
  basePath,
  totalPage,
  currentPage,
}) => {
  const pages = [...Array(2 * width + 1)]
    .map((_, i) => i + currentPage - width)
    .filter((page) => page >= 1 && page <= totalPage)
  if (pages.length < 2 || !pages.includes(currentPage)) return null
  const path = (page: number): string => {
    if (page === 1) return basePath
    return `${basePath}/${page}`.replace(/\/\/+/g, "/")
  }
  return (
    <HStack spacing={3} fontSize={"xl"}>
      {currentPage !== 1 && (
        <Link as={GatsbyLink} to={path(currentPage - 1)}>
          {"<"}
        </Link>
      )}
      {pages.map((page) => {
        if (page === currentPage) {
          return (
            <Box key={page} as={"span"} color={"tint"}>
              {page}
            </Box>
          )
        }
        return (
          <Link key={page} as={GatsbyLink} to={path(page)}>
            {page}
          </Link>
        )
      })}
      {currentPage !== totalPage && (
        <Link as={GatsbyLink} to={path(currentPage + 1)}>
          {">"}
        </Link>
      )}
    </HStack>
  )
}
