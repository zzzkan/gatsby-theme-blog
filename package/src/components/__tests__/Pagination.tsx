import React from "react"
import { render } from "@testing-library/react"
import { Pagination } from "../Pagination"

describe("Pagination component", () => {
  test("snapshot", () => {
    const { asFragment } = render(
      <Pagination basePath={"/"} totalPage={10} currentPage={5} />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  test("should not render anything if totalPage < 2", () => {
    const { container } = render(
      <Pagination basePath={"/"} totalPage={1} currentPage={1} />
    )
    expect(container.firstChild).toBeNull()
  })

  test("should not render anything if currentPage < 1", () => {
    const { container } = render(
      <Pagination basePath={"/"} totalPage={10} currentPage={0} />
    )
    expect(container.firstChild).toBeNull()
  })

  test("should not render '<' if currentPage == 1", () => {
    const { queryByText } = render(
      <Pagination basePath={"/"} totalPage={10} currentPage={1} />
    )
    expect(queryByText("<")).toBeNull()
    expect(queryByText("0")).toBeNull()
    expect(queryByText("1")).not.toHaveAttribute("href", "/")
    expect(queryByText("2")).toHaveAttribute("href", "/2")
    expect(queryByText("3")).toHaveAttribute("href", "/3")
    expect(queryByText("4")).toBeNull()
    expect(queryByText(">")).toHaveAttribute("href", "/2")
  })

  test("should render '<' if currentPage > 1", () => {
    const { queryByText } = render(
      <Pagination basePath={"/"} totalPage={10} currentPage={2} />
    )
    expect(queryByText("<")).toHaveAttribute("href", "/")
    expect(queryByText("0")).toBeNull()
    expect(queryByText("1")).toHaveAttribute("href", "/")
    expect(queryByText("2")).not.toHaveAttribute("href", "/2")
    expect(queryByText("3")).toHaveAttribute("href", "/3")
    expect(queryByText("4")).toHaveAttribute("href", "/4")
    expect(queryByText("5")).toBeNull()
    expect(queryByText(">")).toHaveAttribute("href", "/3")
  })

  test("should render '>' if currentPage < totalPage", () => {
    const { queryByText } = render(
      <Pagination basePath={"/base/"} totalPage={10} currentPage={9} />
    )
    expect(queryByText("<")).toHaveAttribute("href", "/base/8")
    expect(queryByText("6")).toBeNull()
    expect(queryByText("7")).toHaveAttribute("href", "/base/7")
    expect(queryByText("8")).toHaveAttribute("href", "/base/8")
    expect(queryByText("9")).not.toHaveAttribute("href", "/base/9")
    expect(queryByText("10")).toHaveAttribute("href", "/base/10")
    expect(queryByText("11")).toBeNull()
    expect(queryByText(">")).toHaveAttribute("href", "/base/10")
  })

  test("should not render '>' if currentPage == totalPage", () => {
    const { queryByText } = render(
      <Pagination basePath={"/base"} totalPage={10} currentPage={10} />
    )
    expect(queryByText("<")).toHaveAttribute("href", "/base/9")
    expect(queryByText("7")).toBeNull()
    expect(queryByText("8")).toHaveAttribute("href", "/base/8")
    expect(queryByText("9")).toHaveAttribute("href", "/base/9")
    expect(queryByText("10")).not.toHaveAttribute("href", "/base/10")
    expect(queryByText("11")).toBeNull()
    expect(queryByText(">")).toBeNull()
  })

  test("should not render anything if totalPage totalPage", () => {
    const { container } = render(
      <Pagination basePath={"/"} totalPage={10} currentPage={11} />
    )
    expect(container.firstChild).toBeNull()
  })
})
