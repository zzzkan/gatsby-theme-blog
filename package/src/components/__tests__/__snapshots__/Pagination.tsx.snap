// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pagination component > snapshot 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  font-size: xl;
}

.emotion-0>*:not(style)~*:not(style) {
  margin-top: 0px;
  -webkit-margin-end: 0px;
  margin-inline-end: 0px;
  margin-bottom: 0px;
  -webkit-margin-start: 3px;
  margin-inline-start: 3px;
}

.emotion-4 {
  color: tint;
}

<div
    class="chakra-stack emotion-0"
  >
    <a
      aria-label="Move page (/4)"
      class="chakra-link emotion-1"
      href="/4"
    >
      &lt;
    </a>
    <a
      aria-label="Move page (/3)"
      class="chakra-link emotion-1"
      href="/3"
    >
      3
    </a>
    <a
      aria-label="Move page (/4)"
      class="chakra-link emotion-1"
      href="/4"
    >
      4
    </a>
    <span
      class="emotion-4"
    >
      5
    </span>
    <a
      aria-label="Move page (/6)"
      class="chakra-link emotion-1"
      href="/6"
    >
      6
    </a>
    <a
      aria-label="Move page (/7)"
      class="chakra-link emotion-1"
      href="/7"
    >
      7
    </a>
    <a
      aria-label="Move page (/6)"
      class="chakra-link emotion-1"
      href="/6"
    >
      &gt;
    </a>
  </div>
</DocumentFragment>
`;
